### PR TITLE
Cleaner no proxy support

### LINKED ISSUE #1062: Don't call `should_not_be_proxied` on each request
Currently we're calling `should_not_be_proxied` on each request, which is problematic for two reasons...

* It's ignoring the `trust_env`, and relying on an environment variable unilaterally.
* It's preforming an extra chunk of work on each request that we'd rather not do.

### PR DESCRIPTION
**Pull request by tomchristie**
Builds on #1099, and #1098.
Closes #1062.

Refactors our NO_PROXY logic, so that instead of calling `should_not_be_proxied` we instead setup the `proxies` infomation in such a way that it has any exclusions within it, eg...

```
# The following environment...

# ALL_PROXY=http://localhost:1234
# NO_PROXY=example.com

# Is equivalent to...
proxies={"all": httpx.Proxy("http://localhost:1234"), "all://*example.com": None}
```

Note that this PR also introduces wildcard domain prefixing for proxy configuration, since it's required in order to handle `NO_PROXY` equivalents correctly, so we now support...

```python
# Match any domain
proxies={"all://*": httpx.Proxy("http://localhost:1234")}

# Match "example.com" exactly.
proxies={"all://example.com": httpx.Proxy("http://localhost:1234")}

# Match "www.example.com" and "example.com". (But not "wwwexample.com")
proxies={"all://*example.com": httpx.Proxy("http://localhost:1234")}

# Match "www.example.com" (But not "example.com")
proxies={"all://*.example.com": httpx.Proxy("http://localhost:1234")}
```

### CODE REVIEW

File: httpx/_client.py [hunk 1]
```diff
@@ -90,7 +89,7 @@ def _get_proxy_map(
         if proxies is None:
             if trust_env:
                 return {
-                    key: Proxy(url=url)
+                    key: None if url is None else Proxy(url=url)
```
Comment (tomchristie) [inline httpx/_client.py:92]: Previously `get_environment_proxies()` did not return `None` values.
Now it does.

File: httpx/_utils.py [hunk 2]
```diff
@@ -496,10 +512,13 @@ def matches(self, other: "URL") -> bool:
     def priority(self) -> tuple:
         """
         The priority allows URLMatcher instances to be sortable, so that
-        we can match from most specific to least specific.
+        if we can match from most specific to least specific.
```
Comment (florimondmanca) [inline httpx/_utils.py:515]:
```suggestion
        we can match from most specific to least specific.
```

File: tests/test_utils.py [hunk 3]
```diff
@@ -228,75 +227,85 @@ def test_obfuscate_sensitive_headers(headers, output):
...
         ),  # whole-domain matching
         (
             "https://www.example3.com",
-            {"NO_PROXY": "*"},
+            {"ALL_PROXY": "http://localhost:123", "NO_PROXY": "*"},
             True,
         ),  # wildcard * means nothing proxied
     ],
 )
 def test_should_not_be_proxied(url, no_proxy, expected):
```
Comment (florimondmanca) [inline tests/test_utils.py:295]: Should we move this test to `test_proxies.py` now that the `should_not_be_proxied` util is gone?

Comment (tomchristie) [reply tests/test_utils.py:295]: Good call

Comment (tomchristie) [reply tests/test_utils.py]: I've merged these tests into the existing `test_proxies_environ` case now.
Looks much nicer!

**Pull request merged**